### DISCUSSION TITLE
ESP32 | Deep sleep and touch button usage problem

### DISCUSSION THREAD

**Discussion by Elektrofussel** [arendst/Tasmota#12420] [Q&A] [answered]
Hello everybody!  
Since I've been a little more involved with the IOT, more and more ideas have come into my head to make life easier for myself and those around me.  One of my current ideas is self-sufficient garden irrigation.  
Project environment: Allotment garden on the outskirts of a city without a fixed internet connection.  Greenhouse with water barrel and 5V submersible water pump.  There is a 6V solar panel on the roof of the greenhouse.  
How is the system structured?  For the power supply I use a 2500mAh LiPo battery, which is connected to a solar charger.  3 outputs with 5V each are in use from the solar charging unit.  

The first output goes directly to the ESP32, the second output goes to an ESP8266 and the third output goes to a MOSFET "relay".  
The ESP8266 runs with ESP Urna and should act as a hotspot for the ESP32, for later OTA updates and as an interface for user-friendly operation, so that my father-in-law can also adjust the timers and co. 

The core is the ESP32 which has a GPIO 25 defined as a relay, with this output the enable channel of the ESP8266 is set to high.  The GPIO 21 and 22 are for I2C and were connected to a DS3231 for time sync because there is no internet in the garden.  The GPIO pins 36 and 39 are for ADC inputs to check the battery and the voltage applied to the ESP32.  The GPIO 14 is declared as a PWM output to control the MOSFET, which then continuously switches the pump.  
The last GPIO 27 I used is configured as Button_TC to actually bring the ESP32 out of deep sleep.  
Now I come to my real problem. 

The ESP32 is supposed to sleep all day except in the morning and in the evening.  In the morning and in the evening he should wake up at 6:00 a.m. and 7:00 p.m. and water the tomatoes and cucumbers for about 20 minutes, when that is done he should sleep again so as not to empty the battery.  So far so good, everything can be solved using rules and timers using the DeepSleepTimer.  But since I won't be in the garden at 6:00 a.m. and 7:00 p.m. to make optimizations, the ESP32 must also be able to wake up if I want to.  
The manual wake up should be done via the Button_TC, but how?  
So far, the button only switches the relay or the PWM channel, depending on the channel the button was configured.  
My only rule that has worked well so far is Rule1 with „Rule1 on System#Init do TouchNum 1 endon on System#Init do TouchThres 37 endon", but this is only intended for the sensitivity of the button. 
Currently my ESP32 runs with Tasmota 9.4.0, which I compiled via PlatformIO myself to use ADC and some other functions. 
Unfortunately, I can't find a real clue in the documentation about running the ESP32 in DeepSleep. I didn't put GPIO 16 on RST… either, since it is able to wake itself up. 

Has anyone already experience with the DeepSleep function on an ESP32 and can share it with me? 
Perhaps it can really be possible to use the Button_TC as an alarm clock, if so then how? 
In a pinch I will use a switch and define it as DeepSleep in Tasmota and pull it towards Ground to wake up the ESP. 

If you have read everything so far, then I thank you for your attention!  
If the translation into English is bad rather than right, I apologize. I used the classic google translator. 
For more information, suggestions for improvement and solutions, I can be reached at any time via this post! 
I wish you a pleasant day! 😊

![ESP32_Tasmota_PIN-Conf](https://user-images.githubusercontent.com/80831019/122659244-56f25400-d176-11eb-809f-0337186bee71.PNG)

![ESP32_Tasmota_Console-Out](https://user-images.githubusercontent.com/80831019/122659274-ba7c8180-d176-11eb-8fb0-ad8bc130e4f9.PNG)

**Comment by Jason2866** [comment, marked as answer]
Afaik, the Deepsleep feature is not tested with the ESP32.
Seems the way using a button is not implemented for the ESP32
https://github.com/arendst/Tasmota/pull/9048

**Reply by barbudor** [reply]
@Elektrofussel
As of today, Tasmota implementation for DeepSleep is exactly the same as for ESP8266 and only support timer based wake-up for repetitive "wake-up, measure, publish, sleep" mecanism.
None of the wake-up on GPIO mode are supported for now. Just because nobody yet have a use-case for it.

> The last GPIO 27 I used is configured as Button_TC to actually bring the ESP32 out of deep sleep.

Not implemented. If you want to wake up the ESP32 today, you need to reset it.

Not that for timer-based wake-up, the ESP32 doesn't need the external connection of GPIO16 to RST as this is handled internally in ESP32.

**Reply by smhc** [reply]
> None of the wake-up on GPIO mode are supported for now. Just because nobody yet have a use-case for it.

I think it's been requested quite a few times in discussions. A battery operated contact sensor for doors, catflaps, letter boxes, gates, garages etc seems like a pretty practical use case.

I would have thought it'd be relatively simple to make a new command similar to "Restart 9" that takes a gpio pin and level that calls esp_sleep_enable_ext0_wakeup(gpio,level) before doing the deep sleep.

So you could do "Restart 10 4 1" and it would sleep until pin 4 goes high.
You could have rules that handle the wake event similar to how you handle waking from DeepSleepTimer. These rules could check the various gpios to perform whatever action is needed before issuing the same restart command again.
